spec/models/lti_deployment_spec.rb

```diff
@@ -143,4 +143,167 @@
       end
     end
   end
+  describe '#create_or_update_lti_assessment' do
+    let(:course) { create :course }
+    let(:lti_deployment) { create :lti_deployment, course: course }
+    let(:assessment) { create :assignment, course: course }
+    let(:scope) { 'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem' }
+    let(:lti_service_lineitem) { create :lti_service_lineitem, lti_deployment: lti_deployment }
+    before :each do
+      stub_request(:post, Settings.lti.token_endpoint)
+        .with(
+          body: hash_including(
+            { grant_type: 'client_credentials',
+              client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+              scope: scope,
+              client_assertion: /.*/ }
+          ),
+          headers: {
+            Accept: '*/*'
+          }
+        ).to_return(status: :success, body: { access_token: 'access_token',
+                                              scope: scope,
+                                              token_type: 'Bearer',
+                                              expires_in: 3600 }.to_json)
+
+      stub_request(:post, lti_service_lineitem.url)
+        .with(
+          body: hash_including(
+            { label: assessment.description,
+              resourceId: assessment.short_identifier,
+              scoreMaximum: assessment.max_mark.to_f.to_s }
+          ),
+          headers: {
+            Accept: '*/*',
+            'Accept-Encoding': 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
+            Authorization: 'Bearer access_token',
+            'Content-Type': 'application/x-www-form-urlencoded',
+            Host: URI(lti_service_lineitem.url).host,
+            'User-Agent': 'Ruby'
+          }
+        ).to_return(status: :success, body: { id: "#{lti_service_lineitem.url}/1" }.to_json)
+    end
+    it 'creates a new lti line item' do
+      lti_deployment.create_or_update_lti_assessment(assessment)
+      expect(LtiLineItem.count).to eq(1)
+    end
+    it 'sets the correct url' do
+      lti_deployment.create_or_update_lti_assessment(assessment)
+      expect(LtiLineItem.first.lti_line_item_id).to include(lti_service_lineitem.url)
+    end
+  end
+  describe '#create_grades' do
+    let(:course) { create :course }
+    let(:student) { create :student }
+    let(:lti_deployment) { create :lti_deployment, course: course }
+    let!(:assessment) { create :assignment_with_criteria_and_results, course: course }
+
+    let(:scope) { 'https://purl.imsglobal.org/spec/lti-ags/scope/lineitem' }
+    let(:lti_service_lineitem) { create :lti_service_lineitem, lti_deployment: lti_deployment }
+    let(:lti_line_item) { create :lti_line_item, assessment: assessment, lti_deployment: lti_deployment }
+    let(:lti_service_namesrole) { create :lti_service_namesrole, lti_deployment: lti_deployment }
+    let(:status) { 'Active' }
+    before :each do
+      Result.joins(:grouping)
```

Comment:
You don't need this `.joins` because `grouping` is included in the next line. (Same as the tests further down.)